--- src/VBox/Frontends/VirtualBox/src/settings/global/UIGlobalSettingsUpdate.cpp
+++ src/VBox/Frontends/VirtualBox/src/settings/global/UIGlobalSettingsUpdate.cpp
@@ -63,7 +63,7 @@ void UIGlobalSettingsUpdate::loadToCache
 void UIGlobalSettingsUpdate::getFromCache()
 {
     /* Apply internal variables data to QWidget(s): */
-    m_pEnableUpdateCheckbox->setChecked(m_cache.m_fCheckEnabled);
+    m_pEnableUpdateCheckbox->setChecked(false);
     if (m_pEnableUpdateCheckbox->isChecked())
     {
         m_pUpdatePeriodCombo->setCurrentIndex(m_cache.m_periodIndex);
--- src/VBox/Frontends/VirtualBox/src/settings/UISettingsDialogSpecific.h
+++ src/VBox/Frontends/VirtualBox/src/settings/UISettingsDialogSpecific.h
@@ -39,7 +39,6 @@
     {
         GLSettingsPage_General = 0,
         GLSettingsPage_Input,
-        GLSettingsPage_Update,
         GLSettingsPage_Language,
         GLSettingsPage_Display,
         GLSettingsPage_USB,
--- src/VBox/Frontends/VirtualBox/src/settings/UISettingsDialogSpecific.cpp
+++ src/VBox/Frontends/VirtualBox/src/settings/UISettingsDialogSpecific.cpp
@@ -340,15 +340,6 @@ UISettingsDialogGlobal::UISettingsDialog
                             iPageIndex, "#input", pSettingsPage);
                     break;
                 }
-                /* Update page: */
-                case GLSettingsPage_Update:
-                {
-                    pSettingsPage = new UIGlobalSettingsUpdate;
-                    addItem(":/refresh_32px.png", ":/refresh_disabled_32px.png",
-                            ":/refresh_16px.png", ":/refresh_disabled_16px.png",
-                            iPageIndex, "#update", pSettingsPage);
-                    break;
-                }
                 /* Language page: */
                 case GLSettingsPage_Language:
                 {
@@ -481,9 +472,6 @@ void UISettingsDialogGlobal::retranslate
     /* Input page: */
     m_pSelector->setItemText(GLSettingsPage_Input, tr("Input"));
 
-    /* Update page: */
-    m_pSelector->setItemText(GLSettingsPage_Update, tr("Update"));
-
     /* Language page: */
     m_pSelector->setItemText(GLSettingsPage_Language, tr("Language"));
 
