Index: VirtualBox-3.1.4_OSE/src/VBox/Devices/PC/Etherboot-src/core/nic.c
===================================================================
--- VirtualBox-3.1.4_OSE.orig/src/VBox/Devices/PC/Etherboot-src/core/nic.c
+++ VirtualBox-3.1.4_OSE/src/VBox/Devices/PC/Etherboot-src/core/nic.c
@@ -478,6 +478,7 @@ int eth_load(struct dev *dev __unused)
 	}
 	interruptible_sleep(2);		/* lay off the server for a while */
 	longjmp(restart_etherboot, -1);
+	return 0;
 }
 
 
Index: VirtualBox-3.1.4_OSE/src/VBox/Devices/PC/Etherboot-src/core/osloader.c
===================================================================
--- VirtualBox-3.1.4_OSE.orig/src/VBox/Devices/PC/Etherboot-src/core/osloader.c
+++ VirtualBox-3.1.4_OSE/src/VBox/Devices/PC/Etherboot-src/core/osloader.c
@@ -81,6 +81,7 @@ dead_download - Restart etherboot if pro
 **************************************************************************/
 static sector_t dead_download ( unsigned char *data __unused, unsigned int len __unused, int eof __unused) {
         longjmp(restart_etherboot, -2);
+		return 0;
 }
 
 #ifdef	IMAGE_MULTIBOOT
